native/libcst/src/tokenizer/core/mod.rs

```diff
@@ -342,7 +342,20 @@ impl<'t> TokState<'t> {
                         self.missing_nl_before_eof = false;
                         Ok(TokType::Newline)
                     } else {
-                        Ok(TokType::EndMarker)
+                        let hanging_indents = self.indent_stack.len() as i32;
+                        if self.pending_indents == 0 && hanging_indents != 0 {
+                            // We've reached EOF but there are still pending indents not
+                            // accounted for. Flush them out.
+                            self.pending_indents = -hanging_indents;
+                            self.indent_stack.clear();
+                            self.alt_indent_stack.clear();
+                            self.missing_nl_before_eof = false;
+                        }
+                        if let Some(t) = self.process_pending_indents() {
```

Comment:
This seems like it would only generate a single dedent, how does it handle the case where there are multiple levels of indentation to flush?

Reply:
This function is getting called repeatedly and should only produce one token for each invocation. So in effect we'll be calling `process_pending_indents()` repeatedly until it returns `None`, at which point we're done with the input, so we fall down to the else branch on line 356.

Reply:
Ok, that makes more sense. I see why `hanging_indents` is a good name, and how this works conceptually. As long as `pending_indents` is a positive integer we will emit the dedent, and when it is zero we emit the end marker.